### ISSUE TITLE
polish parameter server design doc

### ISSUE THREAD

**Issue by QiJune** [sql-machine-learning/elasticdl#1311]

**Comment by QiJune** [comment]
## Overview

There are commonly two implementations to support distributed training. One is allreduce, the other is parameter server. Sometimes, the model size is out of the memory of a computer, we have to use a distributed storage to save the model. In this scenario, parameter server would be more suitable. The parameter of a model will be sharded among several nodes, which are call PS(parameter server) nodes. Besides, since we use use more PS nodes, the network bandwidth and CPU resources are increased. Thus, the communication between workers and PS, and the optimization in PS will also be speed up. 

In addition, we want to launch machine learning jobs by a Kubernetes supported cluster. Both elastic scheduling and fault-tolerance are very imported in such an environment.  The pods in k8s is scheduled with priority, and could be preempted all the time. If resources become sufficient, we also want to add more resources to running jobs without restart.

In conclusion, a parameter server with scalability and fault-tolerance is highly needed.

**Comment by QiJune** [comment]
## Basic functionalities of PS

PS has three functionalities:

- distributed storage of  model parameters 
- an optimizer, which applies gradients from workers to parameters in PS
- RPC service for workers to pull parameters and push gradients

### KVStore

The typical big parameter of a model could be an embedding table. There could be billions of user ids, item ids in a recommendation system. The size of the table could be larger than 1 TB in some extreme cases. The other parameter could be not such large, usually smaller than 100MB.


The model parameters are sharded among all the PS nodes by a hash function. This is a KVStore instance in a PS node. The KVStore instances of all the PS nodes combine together and become a distributed KVStore, which could be expanded easily to support the large size model.

The KVStore needs to support both embedding table parameter and non-embedding parameter. For the non-embedding parameter, we save it as a tf.Variable in Parameter DB. For the embedding table, we introduce a customized data structure.

The KVStore is the following:

```python
class KVStore(object):
    def __init__(self):
        self.variable_db = {}
        self.embedding_table_db = {}

    def get_parameter(self, name):
        pass

    def get_embedding_vector(self, name, indices):
        pass

    def set_parameter(self, name, value):
        pass

    def set_embedding_vector(self, name, indices, value):
        pass
```

The embedding table is the following:

```python
class EmbeddingTable(object):
    def __init__(self, name, dim, initializer):
        self.name = name
        self.dim = dim
        self.initializer = initializer
        self.vectors = {}

    def get(self, indices):
        pass

    def set(self, indices, value):
        pass
```

Now, we implement the KVStore part with some Python data structures. In the future, we could consider to re-implement the KVStore in C++ to achieve more high performance


### Optimizer

TODO

### RPC service

TODO

**Comment by QiJune** [comment]
## Elastic scheduling and fault-tolerance

### Elastic scheduling

The number of PS node is fixed during the life-cycle of a machine learning jobs. Since the change of PS node number will lead to different parameter sharding strategy. The KVStore instance of each PS node has to be initialized again, which is time-consuming. We only support restart a new job with a different PS node number.

The worker could be added or deleted during the life-cycle of a job at any time. The model parameters are stored at remote PS nodes. In every step, workers pull the latest model for PS, and do forward/backward computation, and push back gradients to PS. There is no local state in a worker.

The initialization of a worker node is the following:

TODO

### Fault-tolerance

TODO

**Comment by workingloong** [comment]
How are variables in model partitioned to PS? The partition strategy design may  need to be added to the document. This [document](https://github.com/tensorflow/community/blob/master/rfcs/20190116-embedding-partitioned-variable.md) introduced the partitioned variable in TF 2.0.

**Comment by QiJune** [comment]
@workingloong Yes, currently, we only shard embedding table to PS nodes under a simple round-robin policy. But some large variable also needs to be sharded to balance pserver's load. We will support it later.

**Comment by mhaoli** [comment]
Closing this issue as we have already finished parameter server design doc.
